emote/optimizers.py

```diff
@@ -0,0 +1,95 @@
+# Adapted from
+# https://github.com/karpathy/minGPT/blob/3ed14b2cec0dfdad3f4b2831f2b4a86d11aef150/mingpt/model.py#L136
+# (MIT license)
+
+from typing import List, Optional, Set, Tuple
+
+import torch
+
+
+def separate_modules_for_weight_decay(
+    network: torch.nn.Module,
+    whitelist_weight_modules: Tuple[torch.nn.Module],
+    blacklist_weight_modules: Tuple[torch.nn.Module],
+    layers_to_exclude: Optional[List[str]] = None,
```

Comment:
If you add `from __future__ import annotations` you can write this as:
```suggestion
    whitelist_weight_modules: tuple[torch.nn.Module, ...],
    blacklist_weight_modules: tuple[torch.nn.Module, ...],
    layers_to_exclude: set[str] | None = None,
```

which is a bit more modern!